pkg/util/time.go

```diff
@@ -98,3 +99,64 @@ func NewDisableableTicker(interval time.Duration) (func(), <-chan time.Time) {
 	tick := time.NewTicker(interval)
 	return func() { tick.Stop() }, tick.C
 }
+
+type TimeRange struct {
+	Start      time.Time
+	End        time.Time
+	Resolution time.Duration
+}
+
+// SplitTimeRangeByResolution splits the given time range into the
+// minimal number of non-overlapping sub-ranges aligned with resolutions.
+// All ranges have inclusive start and end; one millisecond step.
+func SplitTimeRangeByResolution(start, end time.Time, resolutions []time.Duration, fn func(TimeRange)) {
+	if len(resolutions) == 0 {
+		fn(TimeRange{Start: start, End: end})
+		return
+	}
+
+	sort.Slice(resolutions, func(j, i int) bool {
```

Comment:
This could be just me, but the order here (`j` -> `i`) threw me off a bit. If you don't pay attention it looks like resolutions are sorted in an ascending order which is not the case.

Reply:
> This could be just me, but the order here (j -> i) threw me off a bit. If you don't pay attention it looks like resolutions are sorted in an ascending order which is not the case.

I'm pretty sure resolutions are sorted in an ascending order. Swapping `Less`'s args is used e.g. in the standard `sort.Reverse`:
```go
func Reverse(data Interface) Interface {
	return &reverse{data}
}

func (r reverse) Less(i, j int) bool {
	return r.Interface.Less(j, i)
}
```

What caught my eye is that the equality was not strict. Although it should not have any difference in our case, I had no intent – probably I should be more careful with github copilot suggestions :)

Anyways, I changed the order to `i, j` to not confuse anyone

Reply:
Oh, I think I see where the ascending / descending confusion came from. The durations are sorted in descending order, but if viewed as resolutions (which they also are) they are sorted in an ascending order (lower resolution first) :)